### PR TITLE
Add option to show resource usage at termination

### PR DESCRIPTION
**Pull request by rsmmr**

### CODE REVIEW

File: hilti/include/rt/global-state.h [hunk 1]
```diff
@@ -44,6 +46,9 @@ struct GlobalState {
     /** If not zero, `Configuration::abort_on_exception` is disabled. */
     int disable_abort_on_exceptions = 0;
 
+    /** getrusage() at library initialization time. */
+    struct rusage rusage_init;
```
Comment (bbannier) [inline hilti/include/rt/global-state.h:50]: I'd much more prefer if we used a `ResourceUsage` here, if only to ensure that we can consistently collect and compare data we care about, e.g., in `hilti::rt::init` we could just have
```.cpp
rusage_init = rt::resource_usage();
```

File: hilti/src/rt/util.cc [hunk 2]
```diff
@@ -50,13 +51,19 @@ void hilti::rt::abort_with_backtrace() {
 
 void hilti::rt::cannot_be_reached() { hilti::rt::internalError("code is executing that should not be reachable"); }
 
-hilti::rt::MemoryStatistics hilti::rt::memory_statistics() {
-    MemoryStatistics stats;
+hilti::rt::ResourceUsage hilti::rt::resource_usage() {
+    ResourceUsage stats;
 
     struct rusage r;
     getrusage(RUSAGE_SELF, &r);
```
Comment (bbannier) [inline hilti/src/rt/util.cc:58]: This was missing error handling, see my comment on `hilti/src/rt/init.cc`.

File: hilti/src/rt/util.cc [hunk 3]
```diff
@@ -50,13 +51,19 @@ void hilti::rt::abort_with_backtrace() {
...
     auto fibers = detail::Fiber::statistics();
 
+    const auto to_double = [](const auto& x) {
+        return static_cast<double>(x.tv_sec) + static_cast<double>(x.tv_usec) / 1e6;
+    };
```
Comment (bbannier) [inline hilti/src/rt/util.cc:63]: Maybe we could be clearer on types and units here, e.g.,
```suggestion
        const auto to_seconds = [](const timeval& t) {
        return static_cast<double>(t.tv_sec) + static_cast<double>(t.tv_usec) / 1e6;
    };
```

File: hilti/src/rt/tests/util.cc [hunk 4]
```diff
@@ -361,9 +362,11 @@ TEST_CASE("map_tuple") {
     CHECK_EQ(input, std::make_tuple(2u, 2L, std::string("aa")));
 }
 
-TEST_CASE("memory_statistics") {
-    auto ms = memory_statistics();
+TEST_CASE("resource_usage") {
+    auto ms = resource_usage();
 
+    usleep(10000);
```
Comment (bbannier) [inline hilti/src/rt/tests/util.cc:368]: Why is this needed? I'd imagine this could be useful if we'd test for a change in resource usage, but we don't.

File: hilti/src/rt/tests/util.cc [hunk 5]
```diff
@@ -361,9 +362,11 @@ TEST_CASE("map_tuple") {
     CHECK_EQ(input, std::make_tuple(2u, 2L, std::string("aa")));
 }
 
-TEST_CASE("memory_statistics") {
-    auto ms = memory_statistics();
+TEST_CASE("resource_usage") {
+    auto ms = resource_usage();
```
Comment (bbannier) [inline hilti/src/rt/tests/util.cc:366]: Could rename this to `ru` now.

File: spicy/src/rt/driver.cc [hunk 6]
```diff
@@ -70,7 +70,7 @@ void Driver::_debug_stats(size_t current_sessions) {
 
     debug(fmt("sessions: current=%s total=%s", num_sessions, total_sessions));
 
-    auto stats = hilti::rt::memory_statistics();
+    auto stats = hilti::rt::resource_usage();
```
Comment (bbannier) [inline spicy/src/rt/driver.cc:73]: Ditto.

File: spicy/src/rt/driver.cc [hunk 7]
```diff
@@ -51,7 +51,7 @@ void Driver::_debug_stats(const hilti::rt::ValueReference<hilti::rt::Stream>& da
     debug(fmt("input : size-current=%s size-total=%s chunks-cur=%s offset-head=%" PRIu64 " offset-tail=%" PRIu64,
               data_size_cur, data_size_total, data_chunks, data_begin, data_end));
 
-    auto stats = hilti::rt::memory_statistics();
+    auto stats = hilti::rt::resource_usage();
```
Comment (bbannier) [inline spicy/src/rt/driver.cc:54]: It would probably make sense to output the added information on l.60.

This would be simplest and easiest to maintain if we'd provide an `operator<<` for `ResourceUsage` and then just output `debug(fmt("%s", &stats))`.

File: tests/Scripts/canonify-spicy-debug [hunk 8]
```diff
@@ -5,4 +5,5 @@
 cat \
   | sed 's/stream=0x[0-9a-z]\{2,\} /stream=0xXXXXXXXX /g' \
   | sed 's/for stream 0x[0-9a-z]\{2,\}/for stream 0xXXXXXXXX/g' \
-  | sed 's/\(memory *: \).*/\1XXX/g'
+  | sed 's/\(memory *: \).*/\1XXX/g' \
+  | grep -v "^# user_time"
```
Comment (bbannier) [inline tests/Scripts/canonify-spicy-debug:9]: This would omit any time information or the complete resource usage information if you follow my suggestion above.

Maybe we could insyead canonify time information similar to how we handle `memory` (canonify it, and not just drop it).

File: tests/spicy/tools/spicy-build-driver.spicy [hunk 9]
```diff
@@ -1,7 +1,8 @@
 # @TEST-GROUP: no-jit
 # @TEST-EXEC: ${SPICY_BUILD} %INPUT
-# @TEST-EXEC: ${SCRIPTS}/printf '\x00\x00\x27\x10\x2a' | HILTI_DEBUG=spicy ./a.out -p Mini::Test >output
+# @TEST-EXEC: ${SCRIPTS}/printf '\x00\x00\x27\x10\x2a' | HILTI_DEBUG=spicy ./a.out -p Mini::Test -U >output
 # @TEST-EXEC: btest-diff output
+# @TEST-EXEC: grep -q "^# user_time" .stderr
```
Comment (bbannier) [inline tests/spicy/tools/spicy-build-driver.spicy:5]: Might not be needed if we canonify time information instead of dropping it.

File: tests/spicy/tools/spicy-driver-batch.spicy [hunk 10]
```diff
@@ -1,6 +1,7 @@
 # @TEST-EXEC: spicyc -d -j -o test.hlto %INPUT
-# @TEST-EXEC: HILTI_DEBUG=spicy-driver:libspicy spicy-driver -F test.dat test.hlto >output
+# @TEST-EXEC: HILTI_DEBUG=spicy-driver:libspicy spicy-driver -F test.dat -U test.hlto >output
 # @TEST-EXEC: btest-diff output
+# @TEST-EXEC: grep -q "^# user_time" .stderr
```
Comment (bbannier) [inline tests/spicy/tools/spicy-driver-batch.spicy:4]: Ditto.

File: hilti/include/rt/util.h [hunk 11]
```diff
@@ -58,17 +58,19 @@ extern void abort_with_backtrace() __attribute__((noreturn));
...
+    // Note when changing this, update `resource_usage()`.
+    double user_time;       //< user time since runtime initialization
+    double system_time;     //< system time since runtime initialization
     uint64_t memory_heap;   //< current size of heap in bytes
     uint64_t num_fibers;    //< number of fibers currently in use
     uint64_t max_fibers;    //< high-water mark for number of fibers in use
     uint64_t cached_fibers; //< number of fibers currently cached for reuse
 };
 
 /** Returns statistics about the current state of memory allocations. */
```
Comment (bbannier) [inline hilti/include/rt/util.h:72]:
```suggestion
/** Returns statistics about the current resource usage. */
```

File: hilti/include/rt/util.h [hunk 12]
```diff
@@ -58,17 +58,19 @@ extern void abort_with_backtrace() __attribute__((noreturn));
 /** Aborts with an internal error saying we should not be where we are. */
 extern void cannot_be_reached() __attribute__((noreturn));
 
-/** Statistics about the current state of memory allocations. */
-struct MemoryStatistics {
-    // Note when changing this, update `memory_statistics()`.
+/** Statistics about current resource uage. */
```
Comment (bbannier) [inline hilti/include/rt/util.h:61]:
```suggestion
/** Statistics about resource usage. */
```

File: hilti/src/rt/debug-logger.cc [hunk 13]
```diff
@@ -26,24 +26,21 @@ void detail::DebugLogger::print(const std::string& stream, const std::string& ms
...
-        if ( ! out.is_open() )
-            fatalError(fmt("libhilti: cannot open file '%s' for debug output", _path));
-
-        _output = std::move(out);
+        if ( _path == "/dev/stdout" )
+            _output = &std::cout;
+        else if ( _path == "/dev/stderr" )
+            _output = &std::cerr;
+        else {
+            _output_file = std::make_unique<std::ofstream>(_path, std::ios::out | std::ios::trunc);
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply hilti/src/rt/debug-logger.cc:38]: the output concludes with `std::endl`, that should take care of the flush, no?

File: hilti/include/rt/global-state.h (hunk 1 above)
Comment (rsmmr) [reply hilti/include/rt/global-state.h:50]: good idea

File: hilti/src/rt/tests/util.cc (hunk 4 above)
Comment (rsmmr) [reply hilti/src/rt/tests/util.cc:368]: I saw test failures for the new `> 0` check on execution time so far, presumably because the double couldn't represent whatever tiny number of execution had passed at that point.

File: spicy/src/rt/driver.cc (hunk 7 above)
Comment (rsmmr) [reply spicy/src/rt/driver.cc:54]: I was debating about that as well but ended up leaving as it is because I'm not sure that information would be very helpful in that debugging output.

File: tests/spicy/tools/spicy-build-driver.spicy (hunk 9 above)
Comment (rsmmr) [reply tests/spicy/tools/spicy-build-driver.spicy:5]: I'll try the canonification but will still leave this in there as btest-diff is done only for debug builds.
